problem/0002-isolated-integration-tests.md

```diff
@@ -0,0 +1,151 @@
+- Contribution Name: Isolated Integration Tests
+- Implementation Owner: @kleimkuhler, @alpeb
+- Start Date: 2020-03-18
+- Target Date: 2020-04-03
+- RFC PR: [linkerd/rfc#12](https://github.com/linkerd/rfc/pull/12)
+- Linkerd Issue:
+  [linkerd/linkerd2#4190](https://github.com/linkerd/linkerd2/issues/4190)
+- Reviewers: @grampelberg @olix0r
+
+# Summary
+
+[summary]: #summary
+
+Linkerd's integration testing process will provide developers a way to test
+basic Linkerd installations as well as more complex ones, such as
+multi-cluster installations and cluster failure scenarios.
+
+A better UX will be available for running all or one of the integration tests.
+Additionally, running the tests will be less reliant on existing cluster
+state.
+
+With these changes to Linkerd's integration testing process, developers will
+be more equip to run Linkerd tests locally without surprises. Adding new tests
+for features that require more complex cluster conditions will be possible.
+
+# Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+The Linkerd integration test suite is extensive and effective at catching
+bugs, but it does not provide a very positive UX experience to new or existing
+Linkerd developers.
+
+It is also unable to provide many options for testing complex cluster
+scenarios such as multi-cluster installations and cluster failure scenarios.
+
+The are several reasons for these existing problems and each should be
+addressed by the change this RFC is proposing.
+
+### UX
+
+Running specific tests is not currently possible without knowing how the
+internal testing scripts work. This can be better so designed so that
+developers are able to test specific Linkerd integrations without running the
+entire test suite, or looking at the script internals.
+
+After this change, the process for running specific integration tests will be
+more perceptible to developers.
+
+### Constraints
+
+The integration test setup checks require that certain conditions are
+satisfied by the given cluster. A surprising condition is that no pre-existing
+Linkerd installation resource may exist; if it does then it is deleted. These
+conditions exist because the test suite is constrained to a single
+cluster--running each test in its own namespace.
+
+After this change, the test suite should isolate tests by cluster. This will
+allow for each test to require any pre-existing state that is required without
+affecting later runs or existing installations.
```

Comment:
Did you mean each test will setup any initial state it requires?

Reply:
> Did you mean each test will setup any initial state it requires?

Yes and no. My wording here is not very specific about how the state is set because the problem phase does not propose whether tests should programmatically setup their state, or if that state is already existing from running scripts like what currently happens.

It looks like for the time being, tests will still depend on scripts setting up the state since there is not a great option for programmatically doing this, but I didn't want to rule that out in this statement.